# CIA World Factbook project

In this project, we use SQL to explore and analyze data from the CIA World Factbook: https://www.cia.gov/library/publications/the-world-factbook/

The demographic information contained in this database:
 - population (as of 2015)
 - population_growth (the annual population growth rate as percentage)
 - area (the total land and water area)

Connect Jupyter Notebook to the database file

In [4]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

'Connected: None@factbook.db'

Query our database to find out the name of the table and how it looks like

In [5]:
%%sql
SELECT *
  FROM sqlite_master
 WHERE type='table';

Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

Return the first 5 rows of the facts table 

In [6]:
%%sql
SELECT *
  FROM facts
 LIMIT 5

Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

Calculate some summary statistics and look for any outlier countries:
- minimum population
- maximum population
- minimum population growth
- maximum population growth

In [13]:
%%sql
SELECT MIN(population) Min_population,
       MAX(population) Max_population,
       MIN(population_growth) Min_population_growth,
       MAX(population_growth) Max_population_growth 
  FROM facts;

Done.


[(0, 7256490011, 0.0, 4.02)]

Some comments after looking at this summary:
- There is a country with a population of 0
- There is a country with a population of over 7 billion people

We will take a look at these countries without using the specific values.

In [14]:
%%sql
SELECT *
  FROM facts
 WHERE population == (SELECT MIN(population)
                        FROM facts
                     );

Done.


[(250, 'ay', 'Antarctica', None, 280000, None, 0, None, None, None, None)]

In [15]:
%%sql
SELECT *
  FROM facts
 WHERE population == (SELECT MAX(population)
                        FROM facts
                     );

Done.


[(261, 'xx', 'World', None, None, None, 7256490011, 1.08, 18.6, 7.8, None)]

The country with 0 population is Antarctica and there is a row for the whole world that containes the 7.2 billion population. 

Both of these make sense now, but let's recalculate the summary statistics from before, excluding the row for the whole world.

In [16]:
%%sql
SELECT MIN(population) Min_population,
       MAX(population) Max_population,
       MIN(population_growth) Min_population_growth,
       MAX(population_growth) Max_population_growth 
  FROM facts
 WHERE population<7256490011

Done.


[(0, 1367485388, 0.0, 4.02)]

Calculate the average value for population and area, rounded to 2 decimals

In [19]:
%%sql
SELECT ROUND(AVG(population), 2) Average_population, 
       ROUND(AVG(area), 2) Average_area
  FROM facts;

Done.


[(62094928.32, 555093.55)]

Find the countries that are densely populated:
- Countries that have above values for population and
- Below average values for area

In [22]:
%%sql
SELECT name, area, population
  FROM facts
 WHERE population > (SELECT AVG(population) FROM facts) AND
       area < (SELECT AVG(area) FROM facts)
 ORDER BY population

Done.


[('United Kingdom', 243610, 64088222),
 ('Thailand', 513120, 67976405),
 ('Germany', 357022, 80854408),
 ('Vietnam', 331210, 94348835),
 ('Philippines', 300000, 100998376),
 ('Japan', 377915, 126919659),
 ('Bangladesh', 148460, 168957745)]

 What country has the most people?
 What country has the highest growth rate?

%%sql
SELECT name AS Most_populated_country
  FROM facts
 WHERE population == (SELECT MAX(population) FROM facts WHERE name<>'World') 
       

In [29]:
%%sql
SELECT name, population_growth AS Highest_pop_growth
  FROM facts
 WHERE population_growth == (SELECT MAX(population_growth) FROM facts) 

Done.


[('South Sudan', 4.02)]

- Which countries have the highest ratios of water to land? 
- Which countries have more water than land?

In [37]:
%%sql
SELECT name, area_water, area_land
FROM facts


Done.


[('Afghanistan', 0, 652230),
 ('Albania', 1350, 27398),
 ('Algeria', 0, 2381741),
 ('Andorra', 0, 468),
 ('Angola', 0, 1246700),
 ('Antigua and Barbuda', 0, 442),
 ('Argentina', 43710, 2736690),
 ('Armenia', 1540, 28203),
 ('Australia', 58920, 7682300),
 ('Austria', 1426, 82445),
 ('Azerbaijan', 3971, 82629),
 ('Bahamas, The', 3870, 10010),
 ('Bahrain', 0, 760),
 ('Bangladesh', 18290, 130170),
 ('Barbados', 0, 430),
 ('Belarus', 4700, 202900),
 ('Belgium', 250, 30278),
 ('Belize', 160, 22806),
 ('Benin', 2000, 110622),
 ('Bhutan', 0, 38394),
 ('Bolivia', 15280, 1083301),
 ('Bosnia and Herzegovina', 10, 51187),
 ('Botswana', 15000, 566730),
 ('Brazil', 157630, 8358140),
 ('Brunei', 500, 5265),
 ('Bulgaria', 2390, 108489),
 ('Burkina Faso', 400, 273800),
 ('Burma', 23070, 653508),
 ('Burundi', 2150, 25680),
 ('Cambodia', 4520, 176515),
 ('Cameroon', 2730, 472710),
 ('Canada', 891163, 9093507),
 ('Cabo Verde', 0, 4033),
 ('Central African Republic', 0, 622984),
 ('Chad', 24800, 1259200),
 ('Chile', 12290, 743812),
 ('China', 270550, 9326410),
 ('Colombia', 100210, 1038700),
 ('Comoros', 0, 2235),
 ('Congo, Democratic Republic of the', 77810, 2267048),
 ('Congo, Republic of the', 500, 341500),
 ('Costa Rica', 40, 51060),
 ("Cote d'Ivoire", 4460, 318003),
 ('Croatia', 620, 55974),
 ('Cuba', 1040, 109820),
 ('Cyprus', 10, 9241),
 ('Czech Republic', 1620, 77247),
 ('Denmark', 660, 42434),
 ('Djibouti', 20, 23180),
 ('Dominica', 0, 751),
 ('Dominican Republic', 350, 48320),
 ('Ecuador', 6720, 276841),
 ('Egypt', 6000, 995450),
 ('El Salvador', 320, 20721),
 ('Equatorial Guinea', 0, 28051),
 ('Eritrea', 16600, 101000),
 ('Estonia', 2840, 42388),
 ('Ethiopia', 104300, None),
 ('Fiji', 0, 18274),
 ('Finland', 34330, 303815),
 ('France', 3374, 640427),
 ('Gabon', 10000, 257667),
 ('Gambia, The', 1180, 10120),
 ('Georgia', 0, 69700),
 ('Germany', 8350, 348672),
 ('Ghana', 11000, 227533),
 ('Greece', 1310, 130647),
 ('Grenada', 0, 344),
 ('Guatemala', 1730, 107159),
 ('Guinea', 140, 245717),
 ('Guinea-Bissau', 8005, 28120),
 ('Guyana', 18120, 196849),
 ('Haiti', 190, 27560),
 ('Honduras', 200, 111890),
 ('Hungary', 3420, 89608),
 ('Iceland', 2750, 100250),
 ('India', 314070, 2973193),
 ('Indonesia', 93000, 1811569),
 ('Iran', 116600, 1531595),
 ('Iraq', 950, 437367),
 ('Ireland', 1390, 68883),
 ('Israel', 440, 20330),
 ('Italy', 7200, 294140),
 ('Jamaica', 160, 10831),
 ('Japan', 13430, 364485),
 ('Jordan', 540, 88802),
 ('Kazakhstan', 25200, 2699700),
 ('Kenya', 11227, 569140),
 ('Kiribati', 0, 811),
 ('Korea, North', 130, 120408),
 ('Korea, South', 2800, 96920),
 ('Kosovo', 0, 10887),
 ('Kuwait', 0, 17818),
 ('Kyrgyzstan', 8150, 191801),
 ('Laos', 6000, 230800),
 ('Latvia', 2340, 62249),
 ('Lebanon', 170, 10230),
 ('Lesotho', 0, 30355),
 ('Liberia', 15049, 96320),
 ('Libya', 0, 1759540),
 ('Liechtenstein', 0, 160),
 ('Lithuania', 2620, 62680),
 ('Luxembourg', 0, 2586),
 ('Macedonia', 280, 25433),
 ('Madagascar', 5501, 581540),
 ('Malawi', 24404, 94080),
 ('Malaysia', 1190, 328657),
 ('Maldives', 0, 298),
 ('Mali', 20002, 1220190),
 ('Malta', 0, 316),
 ('Marshall Islands', 0, 181),
 ('Mauritania', 0, 1030700),
 ('Mauritius', 10, 2030),
 ('Mexico', 20430, 1943945),
 ('Micronesia, Federated States of', 0, 702),
 ('Moldova', 960, 32891),
 ('Monaco', 0, 2),
 ('Mongolia', 10560, 1553556),
 ('Montenegro', 360, 13452),
 ('Morocco', 250, 446300),
 ('Mozambique', 13000, 786380),
 ('Namibia', 1002, 823290),
 ('Nauru', 0, 21),
 ('Nepal', 3830, 143351),
 ('Netherlands', 7650, 33893),
 ('New Zealand', None, 267710),
 ('Nicaragua', 10380, 119990),
 ('Niger', 300, 1266700),
 ('Nigeria', 13000, 910768),
 ('Norway', 19520, 304282),
 ('Oman', 0, 309500),
 ('Pakistan', 25220, 770875),
 ('Palau', 0, 459),
 ('Panama', 1080, 74340),
 ('Papua New Guinea', 9980, 452860),
 ('Paraguay', 9450, 397302),
 ('Peru', 5220, 1279996),
 ('Philippines', 1830, 298170),
 ('Poland', 8430, 304255),
 ('Portugal', 620, 91470),
 ('Qatar', 0, 11586),
 ('Romania', 8500, 229891),
 ('R

Some values are set to none instead of 0 and we need to change that.

In [42]:
%%sql
UPDATE facts SET area_water=0 WHERE area_water IS NULL


0 rows affected.


[]

In [41]:
%%sql
UPDATE facts SET area_land=0 WHERE area_land IS NULL

15 rows affected.


[]

In [66]:
%%sql
SELECT name, area_water, area_land,
       CAST (area_water AS FLOAT)/CAST (area_land AS FLOAT) Ratio
  FROM facts
 ORDER BY Ratio DESC
 LIMIT 10
 

Done.


[('British Indian Ocean Territory', 54340, 60, 905.6666666666666),
 ('Virgin Islands', 1564, 346, 4.520231213872832),
 ('Puerto Rico', 4921, 8870, 0.5547914317925592),
 ('Bahamas, The', 3870, 10010, 0.3866133866133866),
 ('Guinea-Bissau', 8005, 28120, 0.2846728307254623),
 ('Malawi', 24404, 94080, 0.25939625850340137),
 ('Netherlands', 7650, 33893, 0.22571032366565366),
 ('Uganda', 43938, 197100, 0.22292237442922375),
 ('Eritrea', 16600, 101000, 0.16435643564356436),
 ('Liberia', 15049, 96320, 0.15623961794019933)]

These are the top 10 countries with the highest ratio of water to land: British Indian Ocean Territory, Virgin Islands, Puerto Rico, The Bahamas, Guinea-Bissau, Malawi, Netherlands, Uganda, Eritrea, Liberia

In [67]:
%%sql
SELECT name, area_water, area_land
  FROM facts
 WHERE area_water>area_land

Done.


[('Ethiopia', 104300, 0),
 ('British Indian Ocean Territory', 54340, 60),
 ('Virgin Islands', 1564, 346)]

It looks like here are some wrong data in our dataset..Ethiopia should have land area. Disconsidering that, we only have 2 countries with more water than land: British Indian Ocean Territory and Virgin Islands.

Which countries will add the most people to their population next year?

In [78]:
%%sql
SELECT name, population, population_growth,
       population*population_growth/100 people_added
  FROM facts
  WHERE name<>'World'
 ORDER BY people_added DESC
 LIMIT 6

Done.


[('India', 1251695584, 1.22, 15270686.1248),
 ('China', 1367485388, 0.45, 6153684.246),
 ('Nigeria', 181562056, 2.45, 4448270.372),
 ('Pakistan', 199085847, 1.46, 2906653.3662),
 ('Ethiopia', 99465819, 2.89, 2874562.1691),
 ('Bangladesh', 168957745, 1.6, 2703323.92)]

Top 5 countries that will add the most people to their population next year are: India, China, Nigeria, Pakistan, Ethiopia

Which countries have a higher death rate than birth rate?

In [75]:
%%sql
SELECT name, birth_rate, death_rate
  FROM facts
 WHERE birth_rate<death_rate

Done.


[('Austria', 9.41, 9.42),
 ('Belarus', 10.7, 13.36),
 ('Bosnia and Herzegovina', 8.87, 9.75),
 ('Bulgaria', 8.92, 14.44),
 ('Croatia', 9.45, 12.18),
 ('Czech Republic', 9.63, 10.34),
 ('Estonia', 10.51, 12.4),
 ('Germany', 8.47, 11.42),
 ('Greece', 8.66, 11.09),
 ('Hungary', 9.16, 12.73),
 ('Italy', 8.74, 10.19),
 ('Japan', 7.93, 9.51),
 ('Latvia', 10.0, 14.31),
 ('Lithuania', 10.1, 14.27),
 ('Moldova', 12.0, 12.59),
 ('Monaco', 6.65, 9.24),
 ('Poland', 9.74, 10.19),
 ('Portugal', 9.27, 11.02),
 ('Romania', 9.14, 11.9),
 ('Russia', 11.6, 13.69),
 ('Serbia', 9.08, 13.66),
 ('Slovenia', 8.42, 11.37),
 ('Ukraine', 10.72, 14.46),
 ('Saint Pierre and Miquelon', 7.42, 9.72)]

In [77]:
%%sql
SELECT COUNT(*)
  FROM facts

Done.


[(261,)]

There are 23 countries out of 261 with higher death rate and birth rate.

What countries have the highest population/area ratio?

In [82]:
%%sql
SELECT name, population, area, population/area ratio
  FROM facts
 ORDER BY ratio DESC
 LIMIT 10
 

Done.


[('Macau', 592731, 28, 21168),
 ('Monaco', 30535, 2, 15267),
 ('Singapore', 5674472, 697, 8141),
 ('Hong Kong', 7141106, 1108, 6445),
 ('Gaza Strip', 1869055, 360, 5191),
 ('Gibraltar', 29258, 6, 4876),
 ('Bahrain', 1346613, 760, 1771),
 ('Maldives', 393253, 298, 1319),
 ('Malta', 413965, 316, 1310),
 ('Bermuda', 70196, 54, 1299)]

The most populated countries/area in our 2015 dataset where: Macau, Monaco, Singapore, Hong Kong, Gaza Strip.

# Conclusion

This was a practice and play project to understand the basic SQL syntax and see how basic summary statistics can be done using SQL.

I will definitely explore more of it, but for a first run it seems very straight forward and easy to use.